File: packages/azure/android/src/main/java/com/clarionhq/azure/AzureTypes.kt
package com.clarionhq.azure

/**
 * Internal error type thrown by the Android AzureSession.
 * Mapped to NativeAzureError at the Nitro bridge layer.
 */
internal class AzureError(
  val code: String,
  message: String,
  val recoverable: Boolean = false,
  cause: Throwable? = null,
) : Exception(message, cause)

internal data class AzureConfig(
  val language: String,
  val emitPartials: Boolean,
  val emitAudioLevel: Boolean,
  val audioLevelIntervalMs: Int,
  val subscriptionKey: String,
  val region: String,
  val authToken: String,
  val endpoint: String,
  val outputFormat: String,           // "simple" | "detailed"
  val profanity: String,              // "masked" | "removed" | "raw" | "none"
  val enableSpeakerDiarization: Boolean,
  val autoDetectLanguages: List<String>,
  // P6 additions
  val silenceTimeoutMs: Long,         // 0 = disabled
  val phraseHints: List<String>,      // empty = disabled
  val degradeOnTierMismatch: Boolean, // fall back to non-diarization when refused
)

/** Word-level segment. Azure populates these on detailed output + word-level timestamps. */
internal data class AzureSegment(
  val text: String,
  val startMs: Double,
  val durationMs: Double,
  /** -1 when unknown. */
  val confidence: Double,
  val alternatives: List<String>,
)

internal data class AzureTranscript(
  val id: String,
  val sessionId: String,
  val timestamp: Double,
  val text: String,
  val isFinal: Boolean,
  val language: String,
  /** -1 when unknown. */
  val confidence: Double,
  /** -1 when unknown. */
  val offsetMs: Double,
  /** -1 when unknown. */
  val durationMs: Double,
  /** Empty unless diarization is enabled. */
  val speakerId: String,
  val segments: List<AzureSegment>,
)

internal interface AzureCallbacks {
  fun onState(state: String)
  fun onAudioLevel(rms: Double, peak: Double)
  fun onPartial(transcript: AzureTranscript)
  fun onFinal(transcript: AzureTranscript)
  fun onError(error: AzureError)
  /** `kind` = "started" or "ended". `offsetMs` in ms since session start, or -1 if unknown. */
  fun onSpeechBoundary(kind: String, offsetMs: Double)
}

File: packages/azure/ios/AzureTypes.swift
import Foundation

/// Internal error type thrown by the iOS AzureSession.
/// Mapped to NativeAzureError at the Nitro bridge layer.
internal struct AzureError: Error, CustomStringConvertible {
  let code: String
  let message: String
  let recoverable: Bool

  init(code: String, message: String, recoverable: Bool = false) {
    self.code = code
    self.message = message
    self.recoverable = recoverable
  }

  var description: String { "[\(code)] \(message)" }
}

internal struct AzureConfig {
  let language: String
  let emitPartials: Bool
  let emitAudioLevel: Bool
  let audioLevelIntervalMs: Double

  // Auth — exactly one combination must be present.
  let subscriptionKey: String
  let region: String
  let authToken: String
  let endpoint: String

  let outputFormat: String           // "simple" | "detailed"
  let profanity: String              // "masked" | "removed" | "raw" | "none"
  let enableSpeakerDiarization: Bool
  let autoDetectLanguages: [String]  // empty = disabled

  // P6 additions
  let silenceTimeoutMs: Double       // 0 = disabled
  let phraseHints: [String]          // empty = disabled
  let degradeOnTierMismatch: Bool    // fall back to non-diarization if needed
}

internal struct AzureSegment {
  let text: String
  let startMs: Double
  let durationMs: Double
  let confidence: Double  // -1 when unknown
  let alternatives: [String]
}

internal struct AzureTranscript {
  let id: String
  let sessionId: String
  let timestamp: Double
  let text: String
  let isFinal: Bool
  let language: String
  let confidence: Double  // -1 when unknown
  let offsetMs: Double    // -1 when unknown
  let durationMs: Double  // -1 when unknown
  let speakerId: String   // empty when diarization off
  let segments: [AzureSegment]
}

internal protocol AzureCallbacks: AnyObject {
  func onState(_ state: String)
  func onAudioLevel(rms: Double, peak: Double)
  func onPartial(_ transcript: AzureTranscript)
  func onFinal(_ transcript: AzureTranscript)
  func onError(_ error: AzureError)
  /// `kind` is `"started"` or `"ended"`. `offsetMs` is -1 when unknown.
  func onSpeechBoundary(kind: String, offsetMs: Double)
}
